### PR TITLE
New Resource `alicloud_ots_tunnel`; New Datasource `alicloud_ots_tunnels`; Resource/alicloud_ots_table: Support server side encryption;

### PR DESCRIPTION
**Pull request by okingniko**
New Resource alicloud_ots_tunnel; 
New Datasource alicloud_ots_tunnels; 
Resource/alicloud_ots_table: Support server side encryption;

### CODE REVIEW

File: alicloud/connectivity/client.go [hunk 1]
```diff
@@ -116,6 +117,7 @@ type AliyunClient struct {
 	cloudapiconn                 *cloudapi.Client
 	teaConn                      *rpc.Client
 	tablestoreconnByInstanceName map[string]*tablestore.TableStoreClient
+	otsTunnelConnByInstanceName  map[string]otsTunnel.TunnelClient
```
Comment (okingniko) [inline alicloud/connectivity/client.go:120]: TunnelClient是一个interface

File: website/docs/d/ots_tunnels.html.markdown [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  Provides a list of ots tunnels to the user.
+---
+
+# alicloud\_ots\_tunnels
+
+This data source provides the ots tunnels of the current Alibaba Cloud user.
+
+For information about OTS tunnel and how to use it, see [Tunnel overview](https://www.alibabacloud.com/help/en/tablestore/latest/tunnel-service-overview).
+
+-> **NOTE:** Available in v1.170.0+.
```
Comment (super-eggs) [inline website/docs/d/ots_tunnels.html.markdown:16]: v1.171.0

File: website/docs/r/ots_table.html.markdown [hunk 3]
```diff
@@ -65,6 +67,8 @@ The following arguments are supported:
 * `time_to_live` - (Required) The retention time of data stored in this table (unit: second). The value maximum is 2147483647 and -1 means never expired.
 * `max_version` - (Required) The maximum number of versions stored in this table. The valid value is 1-2147483647.
 * `deviation_cell_version_in_sec` - (Optional, Available in 1.42.0+) The max version offset of the table. The valid value is 1-9223372036854775807. Defaults to 86400.
+* `enable_sse` - (Optional, Available in 1.170.0+) Whether enable OTS server side encryption. Default value is false.
+* `sse_key_type` - (Optional, Available in 1.170.0+) The key type of OTS server side encryption. Only `SSE_KMS_SERVICE` is allowed.
```
Comment (super-eggs) [inline website/docs/r/ots_table.html.markdown:71]: v1.171.0

File: website/docs/r/ots_tunnel.html.markdown [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  Provides an OTS (Open Table Service) tunnel resource.
+---
+
+# alicloud\_ots\_tunnel
+
+Provides an OTS tunnel resource.
+
+For information about OTS tunnel and how to use it, see [Tunnel overview](https://www.alibabacloud.com/help/en/tablestore/latest/tunnel-service-overview).
+
+-> **NOTE:** Available in v1.170.0+.
```
Comment (super-eggs) [inline website/docs/r/ots_tunnel.html.markdown:16]: v1.171.0

File: website/docs/r/ots_tunnel.html.markdown [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+* `expired` - Whether the tunnel has expired.
+* `create_time` - The creation time of the Tunnel.
+* `channels` - The channels of OTS tunnel. Each element contains the following attributes:
+  * `channel_id` - The id of the channel.
+  * `channel_type` - The type of the channel, valid values: `BaseData`, `Stream`.
+  * `channel_status` - The status of the channel, valid values: `WAIT`, `OPEN`, `CLOSING`, `CLOSE`, `TERMINATED`.
+  * `client_id` - The client id of the channel.
+  * `channel_rpo` - The latest consumption time of the channel, unix time in nanosecond.
+
+## Import
```
Comment (super-eggs) [inline website/docs/r/ots_tunnel.html.markdown:93]: `### Timeouts`

File: alicloud/resource_alicloud_ots_tunnel.go [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+func resourceAlicloudOtsTunnel() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAliyunOtsTunnelCreate,
+		Read:   resourceAliyunOtsTunnelRead,
+		Update: resourceAliyunOtsTunnelUpdate,
+		Delete: resourceAliyunOtsTunnelDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
```
Comment (super-eggs) [inline alicloud/resource_alicloud_ots_tunnel.go:23]: Timeouts

File: website/docs/d/ots_tunnels.html.markdown [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+  value = "${data.alicloud_ots_tunnels.tunnels_ds.tunnels.0.id}"
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `instance_name` - The name of OTS instance.
+* `table_name` - The name of OTS table.
```
Comment (super-eggs) [inline website/docs/d/ots_tunnels.html.markdown:38]: instance_name 和 table_name: (Required)

File: alicloud/data_source_alicloud_ots_tunnels.go [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+				Required: true,
+				ForceNew: true,
+			},
+			"ids": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+				Computed: true,
+				ForceNew: true,
+				MinItems: 1,
```
[the remark this answers is not in the document]
Comment (okingniko) [reply alicloud/data_source_alicloud_ots_tunnels.go]: 去掉了MinItems限制。

File: alicloud/resource_alicloud_ots_tunnel.go [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+			},
+			"table_name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"tunnel_name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
```
[the remark this answers is not in the document]
Comment (okingniko) [reply alicloud/resource_alicloud_ots_tunnel.go:42]: 参考 alicloud_ots_instance 和 alicloud_ots_table 里 没有限制。

File: website/docs/d/ots_tunnels.html.markdown (hunk 2 above)
Comment (okingniko) [reply website/docs/d/ots_tunnels.html.markdown]: changed to v1.172.0

File: alicloud/resource_alicloud_ots_tunnel.go (hunk 6 above)
Comment (okingniko) [reply alicloud/resource_alicloud_ots_tunnel.go:27]: Add default timeouts

File: website/docs/r/ots_tunnel.html.markdown (hunk 5 above)
Comment (okingniko) [reply website/docs/r/ots_tunnel.html.markdown:100]: add timeouts

File: website/docs/r/ots_tunnel.html.markdown (hunk 4 above)
Comment (okingniko) [reply website/docs/r/ots_tunnel.html.markdown]: changed to v1.172.0

File: website/docs/d/ots_tunnels.html.markdown (hunk 7 above)
Comment (okingniko) [reply website/docs/d/ots_tunnels.html.markdown]: done

File: alicloud/data_source_alicloud_ots_tunnels.go [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.ValidateRegexp,
+			},
+			"output_file": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			// Computed values
```
Comment (xinmu0426) [inline alicloud/data_source_alicloud_ots_tunnels.go:48]: // Computed values 这个去掉吧

File: alicloud/data_source_alicloud_ots_tunnels_test.go [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+			"names.0":                 CHECKSET,
+			"tunnels.#":               "1",
+			"tunnels.0.instance_name": CHECKSET,
+			"tunnels.0.table_name":    CHECKSET,
+			"tunnels.0.tunnel_name":   CHECKSET,
+			"tunnels.0.tunnel_id":     CHECKSET,
+			"tunnels.0.tunnel_type":   "BaseAndStream",
+			"tunnels.0.tunnel_stage":  "InitBaseDataAndStreamShard",
+			"tunnels.0.create_time":   CHECKSET,
+			"tunnels.0.expired":       "false",
```
Comment (xinmu0426) [inline alicloud/data_source_alicloud_ots_tunnels_test.go:79]: 没有channels的出参校验

Comment (xinmu0426) [inline alicloud/data_source_alicloud_ots_tunnels_test.go:79]: 加上所有的出参校验

Comment (okingniko) [reply alicloud/data_source_alicloud_ots_tunnels_test.go]: done，添加channels检校

Comment (okingniko) [reply alicloud/data_source_alicloud_ots_tunnels_test.go]: done，文档内的出参都覆盖了

File: alicloud/data_source_alicloud_ots_tunnels.go (hunk 10 above)
Comment (okingniko) [reply alicloud/data_source_alicloud_ots_tunnels.go]: done

**Comment by okingniko** [comment]
测试结果：

=== RUN   TestAccAlicloudOtsInstancesDataSource
--- PASS: TestAccAlicloudOtsInstancesDataSource (518.90s)
=== RUN   TestAccAlicloudOtsServiceDataSource
--- PASS: TestAccAlicloudOtsServiceDataSource (6.51s)
=== RUN   TestAccAlicloudOtsTablesDataSource_basic
--- PASS: TestAccAlicloudOtsTablesDataSource_basic (521.29s)
=== RUN   TestAccAlicloudOtsTunnelsDataSource_basic
--- PASS: TestAccAlicloudOtsTunnelsDataSource_basic (546.18s)
=== RUN   TestAccAlicloudOtsInstance_basic
--- PASS: TestAccAlicloudOtsInstance_basic (497.54s)
=== RUN   TestAccAlicloudOtsInstanceHighPerformance
--- PASS: TestAccAlicloudOtsInstanceHighPerformance (504.60s)
=== RUN   TestAccAlicloudOtsInstance_multi
--- PASS: TestAccAlicloudOtsInstance_multi (515.71s)
=== RUN   TestAccAlicloudOtsTable_basic
--- PASS: TestAccAlicloudOtsTable_basic (561.48s)
=== RUN   TestAccAlicloudOtsTable_highPerformance
--- PASS: TestAccAlicloudOtsTable_highPerformance (516.48s)
=== RUN   TestAccAlicloudOtsTable_multi
--- PASS: TestAccAlicloudOtsTable_multi (423.90s)
=== RUN   TestAccAlicloudOtsTable_withTableEncryption
--- PASS: TestAccAlicloudOtsTable_withTableEncryption (512.73s)
=== RUN   TestAccAlicloudOtsTunnel_basic
--- PASS: TestAccAlicloudOtsTunnel_basic (516.13s)

File: alicloud/resource_alicloud_ots_tunnel.go [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+			"tunnel_name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validateOTSTunnelName,
+			},
+			"tunnel_type": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "BaseAndStream",
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_ots_tunnel.go:50]: 这个地方默认值不要直接指定了，如果在不指定的情况下，API有默认值，那么这个地方直接加上 Computed： true 就好了

Comment (xiaozhu36) [reply alicloud/resource_alicloud_ots_tunnel.go:50]: 另外，如果这个参数不支持修改，那么应该设置为 ForceNew 的

**Review by xiaozhu36 — CHANGES_REQUESTED** [review]

**Pull request merged**